### PR TITLE
Support mpd protocol version 0.20

### PR DESCRIPTION
**Pull request by chme**
`mpc` is dropping support for mpd servers with protocol version 0.18 (that is the version forked-daapd supports) - https://www.musicpd.org/news/2018/05/mpc-0-30-released. 

I am sitting some time now on the changes in this pr, but i am unsure how to proceed. The problem is that with version 0.19 the list (and count) commands support one or more `group` parameters. To support these, it is necessary to make the generation of the sql queries more dynamic (generating a group by clause). The resulting sql queries are not optimized for the indexes on the forked-daapd database and thus the performance takes a measurable hit. I found it hard to optimize the queries as this is dependent on the client in use.

Besides mpc i stopped using mpd clients for some time now, so it would be great if i could get some feedback, which clients are of interest to users ... and if it is more important to support the newest mpc or to stay on protocol version 0.18.

### CODE REVIEW

**Comment by LordMyschkin** [comment]
Nice you're still working on improving mpd compatibility although you are not using it - thank's for that.
I use ympd (wich works well as it is, i modified it by commenting out elements that are not supported like cross fading); but if any improvment wold generate compatibility with one of the available IOS Clients (MPD.JS, Soundirok or MPDeluxe, none of which works by now, I guess because of the "listfiles" command), I'd be really happy.
I do not have the impression that MPD-Clients a generally improving, I found only a very few with ongoing development - So im am unsure how to use those new fancy features of new mpd protocols.

File: src/db.c [hunk 1]
```diff
@@ -1523,7 +1525,7 @@ db_build_query_items(struct query_params *qp)
     return NULL;
 
   count = sqlite3_mprintf("SELECT COUNT(*) FROM files f %s;", qc->where);
-  query = sqlite3_mprintf("SELECT f.* FROM files f %s %s %s;", qc->where, qc->order, qc->index);
+  query = sqlite3_mprintf("SELECT f.* FROM files f %s %s %s %s;", qc->where, qc->group, qc->order, qc->index);
```
Comment (ejurgensen) [inline src/db.c]: What does sqlite3_mprintf do if qc->group is null? Print "(null)" in the string? I see qc->order and qc->index get set to empty strings by db_build_query_clause().

**Comment by johnappletree** [comment]
@LordMyschkin You could have a look at [rigelian](https://www.rigelian.net/). By far my favorite MPD client for iOS, and it is in active development.

**Comment by LordMyschkin** [comment]
@johnappletree Thank‘s for the hint, I‘ve installed it - but as long as forked-daapd reports a MPD version < 19.0, it does indeed not seem to work - at least not with full features. And I am somehow ambivalent against that subscription model - although I understand the developers intention to be kept motivated to work on the app, I love the idea of having a free network music player and I have far to many apps I never or seldom use to keep subscriptions to such apps.
But it looks promising - if @chme ever really supports version 19+, maybe I‘ll give it a try...

File: src/db.c (hunk 1 above)
Comment (chme) [reply src/db.c]: I changed it, now qc->group is set to empty string in case no group clause is needed.

**Comment by chme** [comment]
@ejurgensen i think, this is good to be merged or do you have some further remarks?

The performance is imo ok. If it turns out to be too bad for some mpd clients, adding an additional index or implementing some special case to better fit into the existing indexes, can be done in a later pr.

**Comment by ejurgensen** [comment]
Looks good, merged it now!

**Pull request merged**